packages/net/src/lib.rs

```diff
@@ -0,0 +1,42 @@
+mod provider;
+
+use std::ops::Deref;
+use url::Url;
+
+pub use http::Method;
+
+pub use provider::*;
+
+#[cfg(any(feature = "blocking", feature = "non_blocking"))]
+const USER_AGENT: &str = "Mozilla/5.0 (X11; Linux x86_64; rv:60.0) Gecko/20100101 Firefox/81.0";
+
+pub trait NetProvider<T> {
+    fn fetch<H>(&self, url: Url, handler: H)
+    where
+        H: RequestHandler<T>;
+}
+
+impl<T, P, D> NetProvider<T> for D
+where
+    P: NetProvider<T>,
+    D: Deref<Target = P>,
+{
+    fn fetch<H>(&self, url: Url, handler: H)
+    where
+        H: RequestHandler<T>,
+    {
+        self.deref().fetch(url, handler)
+    }
+}
+
+pub trait RequestHandler<T>: Send + Sync + 'static {
```

Comment:
I feel like this abstraction could be simplified by having the Request type/trait contain only plain data. For example, it could contain the url/method and then a plain enum `enum Kind { Css, Image, Svg }`. The Request could then be passed back to the Document along with the response data, and the Document could have a `match` statement to parse the data. I feel like that would be simpler than having the net crate "know" about different resource types. Such a simple type could live in the "traits" crate so that "net" and "dom" don't have to depend on each other.

Reply:
The advantage of my current approach is that e.g. the CSS parsing can be done directly on the fetch thread. But changing to a "the document just gets the data" is pretty simple.

Reply:
I had in mind that the `net` crate would be a tokio executor in a single thread, and that further processing could be pushed onto a separate threadpool. But let's go with your approach for the time being.